File: core-py/src/logwayss_core/crypto.py
from __future__ import annotations

import os
from typing import Tuple
import hashlib
from cryptography.hazmat.primitives.ciphers.aead import AESGCM

IV_SIZE = 12
TAG_SIZE = 16


def derive_key(password: bytes | str, salt: bytes, N: int, r: int, p: int, key_len: int = 32) -> bytes:
    """Derive a key using scrypt with given parameters.

    Args:
        password: raw bytes or UTF-8 string.
        salt: random salt bytes.
        N, r, p: scrypt parameters.
        key_len: desired key length (default 32 for AES-256).
    """
    if isinstance(password, str):
        password = password.encode("utf-8")
    return hashlib.scrypt(password=password, salt=salt, n=N, r=r, p=p, dklen=key_len)


def encrypt(aad: bytes, key: bytes, plaintext: bytes) -> Tuple[bytes, bytes, bytes]:
    """AES-256-GCM encrypt. Returns (iv12, tag16, ciphertext)."""
    iv = os.urandom(IV_SIZE)
    aesgcm = AESGCM(key)
    ct_tag = aesgcm.encrypt(iv, plaintext, aad)
    ciphertext, tag = ct_tag[:-TAG_SIZE], ct_tag[-TAG_SIZE:]
    return iv, tag, ciphertext


def decrypt(aad: bytes, key: bytes, iv: bytes, tag: bytes, ciphertext: bytes) -> bytes:
    """AES-256-GCM decrypt using provided iv and tag."""
    aesgcm = AESGCM(key)
    ct_tag = ciphertext + tag
    return aesgcm.decrypt(iv, ct_tag, aad)

File: core-go/internal/crypto/crypto.go
package crypto

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/rand"
	"fmt"

	"golang.org/x/crypto/scrypt"
)

const (
	IVSize  = 12
	TagSize = 16
)

// GenerateSalt creates a random salt of the given size.
func GenerateSalt(size int) ([]byte, error) {
	salt := make([]byte, size)
	if _, err := rand.Read(salt); err != nil {
		return nil, err
	}
	return salt, nil
}

// DeriveKey derives a key using scrypt with the given parameters.
func DeriveKey(password, salt []byte, N, r, p, keyLen int) ([]byte, error) {
	return scrypt.Key(password, salt, N, r, p, keyLen)
}

// Encrypt performs AES-256-GCM with 12-byte IV and returns iv, tag, ciphertext.
func Encrypt(aad, key, plaintext []byte) (iv, tag, ciphertext []byte, err error) {
	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, nil, nil, err
	}
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return nil, nil, nil, err
	}
	iv = make([]byte, IVSize)
	if _, err = rand.Read(iv); err != nil {
		return nil, nil, nil, err
	}
	out := gcm.Seal(nil, iv, plaintext, aad)
	over := gcm.Overhead()
	if len(out) < over {
		return nil, nil, nil, fmt.Errorf("gcm output too short: %d", len(out))
	}
	ciphertext = make([]byte, len(out)-over)
	copy(ciphertext, out[:len(out)-over])
	tag = make([]byte, over)
	copy(tag, out[len(out)-over:])
	return iv, tag, ciphertext, nil
}

// Decrypt performs AES-256-GCM decryption with provided iv and tag.
func Decrypt(aad, key, iv, tag, ciphertext []byte) ([]byte, error) {
	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, err
	}
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return nil, err
	}
	buf := make([]byte, 0, len(ciphertext)+len(tag))
	buf = append(buf, ciphertext...)
	buf = append(buf, tag...)
	return gcm.Open(nil, iv, buf, aad)
}
